### ISSUE TITLE
netty: The optimization of applying for a minimum of 4096 off-heap memory per request has no effect in non-unaryRequest mode

### ISSUE THREAD

**Issue by jammyMarse** [grpc/grpc-java#11719]
**ClientCallImpl:**
![image](https://github.com/user-attachments/assets/2e54a68a-a407-49d1-a362-39fe361452a0)
**MessageFramer:**
![image](https://github.com/user-attachments/assets/4c6e0c91-3913-46f4-9142-c62ed3deed78)
![image](https://github.com/user-attachments/assets/407fa994-d8d7-4dbb-a60f-f1d689bef9af)
![image](https://github.com/user-attachments/assets/5e878041-212e-4371-9d63-50f8a6e99af8)
**NettyWritableBufferAllocator:**
![image](https://github.com/user-attachments/assets/9c3a3ed7-dc5b-4ca8-a5ef-b8fe7a28fcc0)

**In non-unaryRequest mode, 4096 bytes of off-heap memory will be re-applied for each message**
**If you send small packet requests frequently (for example, 100 bytes each time), a lot of external memory will be wasted, and this 4096 optimization logic will not take effect.**

**Comment by ejona86** [comment]
Yeah, it does seem we would keep allocating the 4k and then only using a small part. `commitToSink()` will clear the buffer and it won't be used across messages.

Are you sending many small messages at once, where we could batch them? Or are you needing to send each of the small messages immediately, without waiting for further small messages?

We could remove the MIN_BUFFER part. I'd have to look further if there are any implications there. For many messages sent back-to-back, it'd be good to provide a corking API (which would avoid calling `flush()` after each message), so that they could be encoded into a single buffer.
